src/main/java/com/google/devtools/build/lib/analysis/ConfiguredTarget.java

```diff
@@ -75,4 +75,7 @@ default String getConfigurationChecksum() {
   default SourceArtifact getSourceArtifact() {
     return null;
   }
+
+  @Nullable
```

Comment:
What is this needed for? Looking at `ToolchainContext`, it's not a particularly lightweight object and this change would make it so that it's kept in RAM permanently as opposed to being discarded after the rule in question is analyzed. It's also called a "Context" which makes it doubly surprising that it's kept in RAM after analysis.

Can you run a memory benchmark on a larg-ish target with C++ and Java toolchain resolution enabled?

Reply:
Reason: the execution transition and toolchain transition both need to know about the toolchain context, and this is part of plumbing that through. The full change is here: https://github.com/katre/bazel/tree/et-05-tc-in-atd

I did the memory benchmarking last week, actually. The short version is that adding this increases memory use for analyzing //gws:gws_and_dev_fileset by slightly less than 1%:

Before: 6036MB (blaze memory reported by used-hea-size-after-gc)
After: 6086MB

One option is to store the UnloadedToolchainContext, not the ResolvedToolchainContext. UTC also implements the ToolchainContext interface, but doesn't have the map from toolchain type to ToolchainInfo (or the list of TemplateVariableInfo).

Reply:
(/me acting as the Guardian Of Memory this time)

I looked at the above fork and it looks like that the the new field in `ConfiguredTarget` is used for the sole purpose of passing to `dependentNodeMap()` in production code? If that's the case, that method in itself does a lot of dependency resolution already, so I was wondering if it would be possible to make callers also do toolchain resolution: technically, both toolchains and dependencies are, well, prerequisites and it feels weird that we cache the resolved values for one of them, but not for the other.

Reply:
Shouldn't the main application be in ConfiguredTargetFunction, which already has access to the context?

I see calls in PostConfiguredTargetFunction and TransitionsOutputFormatterCallback (for query). Are these the motivating dependencies?

Reply:
Part of my motivation was also to remove a lot of hacky `/* toolchainsLabels= */ ImmutableSet.of()` style code from around the codebase.

I can change PostConfiguredTargetFunction to re-compute the toolchain resolution process, but TransitionsOutputFormatterCallback can't do that (it doesn't have a skyframe Environment handy). That means that cquery will end up potentially reporting the wrong transitions (as it won't be able to properly set up the execution and toolchain transitions).

The change to use UnloadedToolchainContext instead of the ToolchainContext in ConfiguredTarget gets the correct functionality and reduces the unbounded part of the memory usage (UnloadedToolchainContext has a String, two PlatformInfo objects, a set of ToolchainTypeInfo objects, and a map from ToolchainTypeInfo to Label). Is that acceptable, memory-wise?

Reply:
When I was working in the area, I found that the `toolchainLabels` argument to `dependentNodeMap()` is a bit weird and the `TOOLCHAIN_DEPENDENCY` type is special-cased almost everywhere (in `DependencyResolver`, in `TransitionsOutputFormatterCallback` and so on). So yeah, some revamping is needed in the area and yes, it's probably a good idea to give `TransitionsOutputFormatterCallback` the actual set of toolchain dependencies.

Do I understand correctly that the difference between `UnloadedToolchainContext` and `ResolvedToolchainContext` is that both of them know the toolchain type -> toolchain label mapping, but the latter also knows the actual providers? 

Well, my sentiment is that the less things we plumb the better. Both for RAM and for code cleanliness. The smallest thing that could possibly work is keeping a list of resolved toolchain labels (or maybe a toolchain type -> label mapping). Is that feasible? Then you save even more RAM than storing an `UnloadedToolchainContext` and both `PostConfiguredTargetFunction` and `TransitionsOutputFormatterCallback` are happy.

Reply:
@gregestren: my understanding is that these call sites are in fact the motivating examples.

Reply:
The portions of the toolchain context that are useful (either currently, or after execution and toolchain transitions land):
-  Labels of resolved toolchain dependencies (this currently is also in the configured target's synthetic `$resolved_toolchains_internal` attribute)
- Label of the execution platform (not available at all currently)
- Label of the target platform (also in the configuration but isn't easily accessible without a lot of plumbing)

So the lowest-memory option is to leave all of the `ImmutableSet<Label> toolchainLabels` in place, and add a `Label executionPlatformLabel` for the execution context, and update all call sites. This continues the current ugly hack but doesn't add anything extra to the ConfiguredTarget.

WDYT?

Reply:
Well, it depends on the numbers... given that you need more than a little list of labels (I was hoping that's not the case), I would actually now prefer putting in an `UnloadedToolchainContext` if the difference in memory use is not that dramatic. A consolation prize for having to add that would be that then it'd be possible to remove the synthetic `$resolved_toolchains_internal` attribute (how is that in `ConfiguredTarget` anyway? I thought that it's pretty much just a bag of providers), thus winning back some RAM.

It's fine to lose a bit of RAM, but 1% is a little extreme.

Why do you need a lot of plumbing for the target platform? I thought that the configuration is available everwhere. `TransitionsOutputFormatterCallback` is a little dicey, but `SkyframeExecutor` apparently provides an extra special hook just for that:

https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/query2/TransitionsOutputFormatterCallback.java#L109

And it apparently does its own Skyframe evaluation, which is good and bad: bad because I didn't expect Skyframe evaluation to happen at that place, good because then doing Skyframe evaluation for toolchain resolution is on the table, too...

Reply:
Oh, wow, didn't even notice the executor there, although it doesn't (for good reason) expose generic "look up this SkyKey" functionality. I guess re-doing the toolchain resolution process might be feasible, but would require putting the entire toolchain resolution system behind a new skyfunction.

Re: `$resolved_toolchains_internal`: This is how the actual depended-on toolchains are passed back to the configured target: see [ConfiguredTargetFactory.transformPrerequisiteMap](https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/ConfiguredTargetFactory.java;l=452?q=ConfiguredTargetFactory&ss=bazel). Now that you've updated how dependency resolution works, we might be able to remove it and just pass the proper dependencies into [UnloadedToolchainContext.load](https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/UnloadedToolchainContext.java;l=78?q=UnloadedToolchainContext) directly. I'll explore that in a separate patch (but it may make the TransitionsOutputFormatterCallback not recognize toolchain dependencies. I don't know if that's good or bad).

Reply:
..but is `$resolved_toolchains_internal` kept in the `ConfiguredTarget` instance? Sure, it's needed during its analysis just like `RuleContext` is, but I thought it was then abandoned (I'm on a plane so my code discovery abilities are severely limited...) I don't see any immediately suspicious field in `RuleConfiguredTarget`.

My overarching desire is not to regress on RAM unless there is no alternative. Aside from that, I'd be happy with many options (although I admit I do love nice code!)

Reply:
Looking further, I don't think `$resolved_toolchains_internal` is kept in the ConfiguredTarget instance, so that is definitely a memory regression to be avoided.

Reply:
Is it not kept in the ConfiguredTarget instance to supply https://github.com/bazelbuild/bazel/blob/b3b3e8bb7320ef76dfec97f66284e0a9f7eda2d2/src/main/java/com/google/devtools/build/lib/query2/TransitionsOutputFormatterCallback.java#L124-L128 ? Or you mean it would no longer be necessary if this is refactored to avoid that dependency?

Reply:
I like the idea of exploring a refactoring of how toolchains relate to DependencyResolver / dependency resolution before following through on this change. But it's not clear to me how easy it is to un-special-case them since there are some fundamental ways (maybe?) that toolchain dependencies really *are* special.

Today in DependencyResolver I think the only integration is to set a special TOOLCHAIN_DEPENDENCY on the toolchain labels with a host configuration. We could easily set that directly outside DependencyResolver, or as a special case at the beginning of dependentNodeMap without getting further littered through the class.. But then, @lberki would that get in the way of your intentions of creating distinct phases for partially resolved vs. fully resolve dependencies?

And with John's current changes there's a new role for the toolchain context: providing the execution platform. It's possible to just pass the execution platform directly similar to how we're currently passing in the host config. So I think it's still possible to completely disentangle the toolchain context and DependencyResolver. But I'm not sure that puts us in a better state, nor makes the broader issue of special-casing toolchain dependencies less deep?

Reply:
Anyway, this is all getting complicated. To recap, at core:

- PostConfiguredTarget can just re-call toolchain resolution if it has the unloaded context, which it would have to get from the ConfiguredTarget. This should be quick since that should already be cached in Skyframe. Maybe there's an opportunity to consolidate that same lookup code with other places that resolve toolchains so we're not duplicating a bunch of logic?

- TransitionsOutputFormatterCallback just fits poorly here. It has Skyframe access but through a different interface, which is already an awkward compromise since the current SkyframeExecutor.getConfiguration() is replicating similar logic called from ConfiguredTargetFunction but in a "different" way. Note this is also used for getting the configurations for top-level targets. This is a necessary evil and expanding this with yet another variation for toolchain resolution would be an unfortunate design direction.

But... does TransitionsOutputFormatterCallback actually need to resolve toolchains? If we just pass it the exec platform label via a ConfiguredTarget, I think that's all it needs to function the rest of the way. It doesn't actually *apply* transitions - it just spits out the transition objects. So it should be possible to avoid any extra new Skyframe evaluation needs here.

Reply:
I am going to explore the change where I revert most of this, and just pass the execution platform label directly down the call chain. I'll send it when I have some time.